eth/downloader/skeleton.go

```diff
@@ -1023,6 +1049,54 @@ func (s *skeleton) processResponse(res *headerResponse) (linked bool, merged boo
 	return linked, merged
 }
 
+// cleanStales removes previously synced beacon headers that have become stale
+// due to the downloader backfilling past the tracked tail.
+func (s *skeleton) cleanStales(filled *types.Header) error {
+	number := filled.Number.Uint64()
+	log.Trace("Cleaning stale beacon headers", "filled", number, "hash", filled.Hash())
+
+	// If the filled header is below the linked subchain, something's
+	// corrupted internally. Report and error and refuse to do anything.
+	if number < s.progress.Subchains[0].Tail {
```

Comment:
Actually it can happen in practice.

- We start the beacon sync
- Fetched 10 skeleton headers
- Pivot is picked as HEAD-64, which is in the local chain, less than subchain[0].Tail
- Origin is reset to Pivot-1
- The local chain is reset to Pivot-1, HeadFast=Pivot-1
- And somehow the snap sync is aborted

Reply:
What I meant is: after downloading the beacon chain(only 10 headers) and then we spin up the downloader for bodies and states syncing. Pivot is picked at this time here https://github.com/ethereum/go-ethereum/blob/master/eth/downloader/downloader.go#L483

Because the safety distance for pivot point and head is 64, so pivot point can be recapped to HEAD-64 which is actually in the local header chain instead of the downloaded beacon skeleton.

Reply:
What I meant is: after downloading the beacon chain(only 10 headers) and then we spin up the downloader for bodies and states syncing. Pivot is picked at this time here https://github.com/ethereum/go-ethereum/blob/master/eth/downloader/downloader.go#L483

Because the safety distance for pivot point and head is 64, so pivot point can be recapped to HEAD-64 which is actually in the local header chain instead of the downloaded beacon skeleton.

Reply:
And perhaps another corner case is: after downloading the beacon chain, the sync is aborted for some reasons without inserting any chain segment(the HeadFastBlock is still tail-1).
Maybe we should lower the log level to Info in this case? Although in practice the possibility is really low.

Reply:
Ah, now I see what you meant. Still, isn't that a very rare case? The only scenario I see this happening is syncing during the transition where legacy sync downloaded headers up to the merge block and then when the beacon sync starts, there's not yet 64 blocks in post-merge world. That's kind of 15 minutes in total where this can trigger naturally. Would seem that that's not worth to lower the log level for, but maybe there's some other case?

Reply:
Yup, so let's keep it. Since in this case nothing to delete actually, I think it's fine.